Add downloads factories

jaap3 opened the pull request:
Recreate the downloads section by importing data from the python.org API

Commit:
Add downloads factories

Recreate the downloads section by importing data from the python.org API

berkerpeksag commented on downloads/factories.py:
```diff
@@ -0,0 +1,112 @@
+
+
+class APISession(requests.Session):
+    base_url = 'https://www.python.org/api/v1/'
```
The new API will be deployed to production on this Friday. I would like to use v2 here. If you have some time before Friday, you can try it out at https://staging.python.org/api/v2/

Commit:
Use v2 API

* Responses don't have the 'objects' key
* Responses are not paginated
* Endpoints don't support the 'limit' argument

Commit:
Raise an exception when the a response indicates a client or server error

Commit:
Don't use bulk_create

It only works on a clean database or `create_initial_data` is called with the `--flush` flag

Commit:
Add Accept header

berkerpeksag commented on downloads/factories.py:
```diff
@@ -0,0 +1,120 @@
+    """
+    Get the ID of an object by extracting it from the resource uri.
+    """
+    return (obj.pop(key, '') or '').rstrip('/').rpartition('/')[-1]
```
Do we need the `or ''` part?

jaap3 replied:
I've changed the implementation slightly, should be clearer now

berkerpeksag commented on downloads/factories.py:
```diff
@@ -0,0 +1,120 @@
+
+    """
+    objects = {
+        'oss': {},
```
Typo? os -> oses

jaap3 replied:
yeah, I didn't know how to pluralize OS :)

berkerpeksag commented on downloads/factories.py:
```diff
@@ -0,0 +1,120 @@
+        django_get_or_create = ('slug',)
+
+    creator = factory.SubFactory(UserFactory)
+    is_published = True
```
Could you add `is_latest` as well? We need it to make `update_supernav()` work. https://github.com/python/pythondotorg/commit/5ad3ebf26ff09848557d85587afdf223d57ce20c should expose it in both APIs.

jaap3 replied:
There's no need to add `is_latest` to the factory, if it's on the model then it's also on the factory. So in this case `is_latest` will default to `False`. In `initial_data` the release is created like this `ReleaseFactory(**obj)`, so all fields that are available in the API are set on the newly created instance. So this should already set `is_latest` correctly.

berkerpeksag replied:
Ah, yes, I missed you've added `is_published = True` because its default value is `False`. Good point!

berkerpeksag commented on downloads/factories.py:
```diff
@@ -0,0 +1,120 @@
+        super().__init__(*args, **kwargs)
+        self.headers.update(
+            {
+                "Accept": "application/json",
```
Style nit: Please use single quotes everywhere.

berkerpeksag commented on downloads/factories.py:
```diff
@@ -0,0 +1,120 @@
+    """
+    Create the data for the downloads section by importing
+    it from the python.org API.
+
```
Style nit: Newline can be dropped.

berkerpeksag commented on downloads/factories.py:
```diff
@@ -0,0 +1,120 @@
+
+    # Create all the releases
+    for key, obj in objects['releases'].items():
+        obj.pop('release_page')  # Ignore release pages
```
Did you skip adding release pages for now or did you get blocked by something? I think we can add a `# TODO` comment if it takes too long to implement it.

jaap3 replied:
I think I got blocked on something, but I cannot remember what. I'll add a `TODO` for now

berkerpeksag replied:
Great! Feel free to file an issue if you remember what was the blocker.

berkerpeksag reviewed (commented):
This is great! I needed this when I was working on #1370 and I noticed that I didn't expose the `is_latest` field.

Commit:
Clarify _get_id a bit

Commit:
s/oss/oses

Commit:
Add TODO

berkerpeksag committed:
some very minor style changes

berkerpeksag reviewed (approved):
LGTM! I already tested this last week and will merge once Travis is happy.
